### DISCUSSION TITLE
GA4 dest flattens events params

### DISCUSSION THREAD

**Discussion by Akronae** [rudderlabs/rudder-server#3069] [Q&A]
My JS source receives events with params of this form
```json
{
  "properties": {
    "ecommerce": {
      "items": [
        {
          "currency": "€",
          "item_area": "112",
          "item_category": "SHARED",
          "item_id": "63da914ae0de4e001657d732",
          "item_name": "Coliving - 112m²  - Lyon 69006",
          "item_number_of_rooms": 7,
          "item_reference": "530",
          "item_variant": "Meublé",
          "price": "530"
        }
      ]
    }
  }
}
```

But Rudder flattens it this way on the GA4 destination
```json
{
  "events": [
    {
      "name": "view_item",
      "params": {
        "session_id": "1678093736",
        "engagement_time_msec": 1,
        "ecommerce_items_0_price": "677",
        "ecommerce_items_0_item_id": "63da9337e0de4e001658179f",
        "ecommerce_items_0_currency": "€",
        "ecommerce_items_0_item_area": "67",
        "ecommerce_items_0_item_name": "Coliving - 67m²  - Toulouse 31000",
        "ecommerce_items_0_item_variant": "furnished",
        "ecommerce_items_0_item_category": "SHARED",
        "ecommerce_items_0_item_reference": "677",
        "ecommerce_items_0_item_number_of_rooms": 3
      }
    }
  ]
}
```

I need to keep the original form in order to work with GTM, how can we disable the params flattening?

**Comment by gitcommitshow** [comment]
Are you using GA4 in device mode or cloud mode?

I can see that [GA4 doc for device mode](https://www.rudderstack.com/docs/destinations/streaming-destinations/google-analytics-4/google-analytics-4-device-mode/#track) mentions

> RudderStack SDK sends the track event name and any properties as custom properties to Google Analytics 4.
> RudderStack SDK flattens the properties that are nested and not standard Google Analytics 4 properties before sending them to Google Analytics.

**Reply by Akronae** [reply]
I have `Use device-mode to send events` checked on my GA4 destination.  

From which source did you conclude that
> nested [properties are] not standard Google Analytics 4 properties  

We are implementing a tagging plan from a 3rd party company, so we would need to justify to them why they should change their property naming 😅.

**Reply by Akronae** [reply]
@gitcommitshow is there a way we could build a plugin/transformation to get around this issue? 😃

**Reply by gitcommitshow** [reply]
> @gitcommitshow is there a way we could build a plugin/transformation to get around this issue? 😃 

* This is a limitation at GA4 end, so as long as you want to send events to GA4, we have to follow their spec. 
* But if possible for your use case, you can think of replacing GA4 with your own tooling(i.e. send event data to your warehouse and then analyze yourself using dbt or other tool). RudderStack has event level [JSON column data support](https://www.rudderstack.com/docs/destinations/warehouse-destinations/json-column-support/) for track events. As you said, you'll be using the transformations to achieve this.

**Reply by Akronae** [reply]
> > @gitcommitshow is there a way we could build a plugin/transformation to get around this issue? 😃
> 
> * This is a limitation at GA4 end, so as long as you want to send events to GA4, we have to follow their spec.
> * But if possible for your use case, you can think of replacing GA4 with your own tooling(i.e. send event data to your warehouse and then analyze yourself using dbt or other tool). RudderStack has event level [JSON column data support](https://www.rudderstack.com/docs/destinations/warehouse-destinations/json-column-support/) for track events. As you said, you'll be using the transformations to achieve this.

I could not find any documentation about that limitation on GA4 end, would you have a link? The 3rd party company that want us to use nested params already works with GA4 using this structure and it seems to work fine for them and their customers 🤔

**Reply by gitcommitshow** [reply]
I see. There's definitely a lack of GA4 documentation on this topic. This experience (GA4 not accepting nested params) had been confirmed repeatedly by multiple people.
But if it is working for you, I must check if we missed something in the implementation.
One example I'd want to try is by changing `ecommerce_items_0_price` to `ecommerce_items.price`, will that be parsed as correct nested json by GA4?

**Reply by gitcommitshow** [reply]
@Akronae do you mind sharing a ga4 screenshot showing how this nested data looks in the dashboard?

**Reply by Akronae** [reply]
@gitcommitshow I'm going to ask to have a meetting with the guys in charge of analytics. I was also wondering how nested params would show up in the UI but aparently you can just "select objects" 🤔
Getting back to you soon!
